### ISSUE TITLE
PKCE is not supported

### ISSUE THREAD

**Issue by wilgie** [go-oauth2/oauth2#161]

**Comment by Ekliptor** [comment]
agreed. This would be really cool for react apps without a server (only static files). See https://oauth.net/2/pkce/

**Comment by maximkosov** [comment]
Authentication for mobile apps also require PKCE. Is there any plans to support it?

**Comment by jarlandre** [comment]
authorisation code flow PKCE without passing client_secret is hard to get working. Because null or empty client_secret will cause the logic to not check client_credentials secret. So you cannot use both auth code flow with PKCE and client credentials grant on the same server, because it will allow to get access token for free without passing client secret to token endpoint.

**Comment by jarlandre** [comment]
So to actually fix this problem, where we want to use auth code flow with PKCE without passing client_secret in post body, is to use different functions for getting access token for auth code flow and client credentials

**Comment by jarlandre** [comment]
ok, i can "fix" this on my end by implementing this method and setting it on the server
```go
	ClientScopeHandler func(tgr *oauth2.TokenGenerateRequest) (allowed bool, err error)
```
but i dont feel like it should be a programmer choice.

**Comment by jarlandre** [comment]
made a PR here to fix the issue with client_secret being required in auth code flow with PKCE

https://github.com/go-oauth2/oauth2/pull/230
